testing/architecture-test/src/test/java/org/gradle/architecture/test/PlatformBoundariesTest.java

```diff
@@ -0,0 +1,157 @@
+                    return false;
+                }
+                return allPlatformDirs.stream().anyMatch(classFile::startsWith);
+            }
+        };
+    }
+
+    private static ArchCondition<? super JavaClass> importClassesOnlyFromAllowedPlatforms() {
+        return new ArchCondition<JavaClass>("import classes only from allowed platforms") {
+            @Override
+            public void check(JavaClass ownJavaClass, ConditionEvents events) {
+                Path ownClassFile = getClassFile(ownJavaClass);
+                if (ownClassFile == null) {
+                    return;
+                }
+                String ownPlatformName = platformNameOf(ownJavaClass, ownClassFile);
+                for (JavaAccess<?> access : ownJavaClass.getAccessesFromSelf()) {
+                    JavaClass targetJavaClass = access.getTargetOwner();
+                    Path targetClassFile = getClassFile(targetJavaClass);
+                    if (targetClassFile != null && allPlatformDirs.stream().anyMatch(targetClassFile::startsWith)) {
+                        boolean conditionSatisfied = allowedDirsByPlatform.get(ownPlatformName).stream().anyMatch(targetClassFile::startsWith);
+                        String targetPlatformName = platformNameOf(targetJavaClass, targetClassFile);
+                        String message;
+                        if (conditionSatisfied) {
+                            message = String.format("Class '%s' from platform '%s' does not import any class from forbidden platforms", ownJavaClass.getName(), ownPlatformName);
+                        } else {
+                            message = String.format("Class '%s' from platform '%s' imports class '%s' from a forbidden platform '%s'", ownJavaClass.getName(), ownPlatformName, targetJavaClass.getName(), targetPlatformName);
+                        }
+                        events.add(new SimpleConditionEvent(
+                            ownJavaClass,
+                            conditionSatisfied,
+                            message
+                        ));
+                    }
+                }
+            }
+        };
+    }
+
+    private static Stream<Path> resolvePlatformDirs(Path basePath, List<String> dirNames) {
+        return dirNames.stream().map(basePath::resolve);
+    }
+
+    private static String platformNameOf(JavaClass javaClass, Path classFile) {
+        return dirsByPlatform.entrySet().stream().filter(e -> e.getValue().stream().anyMatch(classFile::startsWith)).map(Map.Entry::getKey).findFirst()
+            .orElseThrow(() -> new IllegalStateException("Could not find platform for class " + javaClass.getName()));
+    }
+
+    @Nullable
+    private static Path getClassFile(JavaClass javaClass) {
+        Class<?> reflectedClass = safeReflect(javaClass);
+        if (reflectedClass == null) {
+            return null;
+        }
+        CodeSource codeSource = reflectedClass.getProtectionDomain().getCodeSource();
```

Comment:
💅 I feel that there's a hidden assumption about the classpath layout, like "the jar/class directory has to be in the build directory of the project". It would be great to have it spelled out explicitly as a comment nearby.